internal/pool/session.go

```
@@ -0,0 +1,523 @@
+	if s.client != nil {
+		s.client.Close()
+		s.client = nil
+	}
+	if s.conn != nil {
+		s.conn.Close()
+		s.conn = nil
+	}
+}
+
+func (s *Session) State() SessionState {
+	state := s.state.Load().(SessionState)
+	return state
+}
+
+func (s *Session) markState(update func(*SessionState)) {
+	current := s.state.Load().(SessionState)
+	update(&current)
+	current.LastUpdated = time.Now()
+	s.state.Store(current)
+}
+
+func (s *Session) reconnect(ctx context.Context) error {
+	s.markState(func(st *SessionState) {
+		st.Phase = WarmupPhaseReconnecting
+		st.Paused = false
+		st.Connected = false
+		st.LastError = ""
+	})
+
+	conn, err := s.dialer.Dial(ctx)
+	if err != nil {
+		s.markState(func(st *SessionState) {
+			st.Phase = WarmupPhaseError
+			st.LastError = err.Error()
+			st.Connected = false
+		})
+		return err
+	}
+
+	s.mu.Lock()
+	if s.client != nil {
+		s.client.Close()
+	}
+	if s.conn != nil {
+		s.conn.Close()
+	}
+	s.conn = conn.RawConn
+	s.client = conn.Client
+	s.mu.Unlock()
+
+	s.markState(func(st *SessionState) {
+		st.Phase = WarmupPhaseContinuous
+		st.LastError = ""
+		st.Connected = true
+	})
+	return nil
+}
+
+func (s *Session) warmupLoop(ctx context.Context) {
+	defer s.warmupWG.Done()
+	for {
+		if s.isClosed() {
+			return
+		}
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		client := s.currentClient()
+		if client == nil {
+			if err := s.reconnect(ctx); err != nil {
+				select {
+				case <-time.After(500 * time.Millisecond):
+				case <-ctx.Done():
+					return
+				}
+				continue
+			}
+			client = s.currentClient()
+			if client == nil {
+				continue
+			}
+		}
+
+		if err := s.runWarmupOnce(ctx, client); err != nil {
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return
+			}
+			s.markState(func(st *SessionState) {
+				st.Phase = WarmupPhaseError
+				st.LastError = err.Error()
+				st.Connected = false
+			})
+			s.reconnect(ctx)
+			continue
+		}
+	}
+}
```

[major] Falling back to  hides a broken invariant. The config layer already guarantees a positive per-connection bit rate, so the internal warm-up logic should never need to defend against zero/negative throughput. Silently throttling to 1 B/s will mask configuration bugs and keep the connection permanently under-warmed. Please drop the fallback and instead tighten validation in  (e.g. ensure the derived per-connection bit rate is large enough that  stays ≥1) so the core logic can rely on a strict invariant and, if violated, fail loudly.